crates/client/sync/src/commitments/lib.rs

```diff
@@ -146,109 +162,130 @@ where
 /// # Returns
 ///
 /// The updated state root as a `Felt252Wrapper`.
-pub async fn update_state_root<B: BlockT>(
+pub fn update_state_root<B: BlockT>(
     csd: CommitmentStateDiff,
-    bonsai_dbs: BonsaiDbs<B>,
-) -> anyhow::Result<Felt252Wrapper> {
-    let arc_csd = Arc::new(csd);
-    let arc_bonsai_dbs = Arc::new(bonsai_dbs);
-
-    let contract_trie_root = contract_trie_root(Arc::clone(&arc_csd), Arc::clone(&arc_bonsai_dbs)).await?;
-
-    let class_trie_root = class_trie_root(Arc::clone(&arc_csd), Arc::clone(&arc_bonsai_dbs))?;
-
-    let state_root = calculate_state_root::<PoseidonHasher>(contract_trie_root, class_trie_root);
-
-    Ok(state_root)
+    overrides: Arc<OverrideHandle<Block<Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>,
+    bonsai_contract: Arc<Mutex<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>>,
+    bonsai_class: Arc<Mutex<BonsaiStorage<BasicId, BonsaiDb<B>, Poseidon>>>,
+    block_number: u64,
+    substrate_block_hash: Option<H256>,
+) -> Felt252Wrapper {
+    // Update contract and its storage tries
+    let contract_trie_root =
```

Comment:
pass the mutex as a parameter in order to lock for the shortest possible time in the child function

Reply:
This really isn't necessary since:

1. each mutex is **only** used in the context of the function it's lock is being passed to.
2. `verify_l2` is never called asynchronously, only the functions inside are to be parallelized.

The use of mutex here is just to allow the madara backend to access the bonsai dbs in a thread safe-context, such as during RPC calls where they might need to be accessed to retrieve hashes.

Reply:
I hadn't seen the for loop in the contract_trie_root function which would force lock and unlock the mutex at each iteration